### DISCUSSION TITLE
Cannot compile iTwin Viewer example

### DISCUSSION THREAD

**Discussion by ksurell** [iTwin/itwinjs-core#2500] [Q&A] [answered]
Hi,
I'm new to this but managed to set up the environment and compile the iTwin viewer example app. The first time.
Now, when I try to create a new app I get a compile error when trying to run the app (I know I can change the app name!).
There are also a lot of warning about depracated packages during the npx create-react-app phase.

Any ideas?!

/Krister 

**Failed to compile**
multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js ./node_modules/react-dev-utils/webpackHotDevClient.js ./src/index.tsx
Module not found: Cannot find module 'C:\Users\krist\your-app-name\node_modules\react-refresh\index.js'. Please verify that the package.json has a valid "main" entry

**Comment by aruniverse** [comment, marked as answer]
Thanks for reporting this, we'll fix this asap.

For now, you should be able to get around this issue by pinning down the version of @bentley/react-scripts in your package.json to 4.0.3 like so:

package.json
![image](https://user-images.githubusercontent.com/11051042/137380217-6384d0d1-31b6-4057-b919-1667baeb22fe.png)

**Reply by aruniverse** [reply]
There should be an updated version 4.0.5 available in the next day or so with the fix.

**Reply by ksurell** [reply]
I just tried this but get the same error.
This is the info in my package.json:

{
  "name": "boo-1-483",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@bentley/bentleyjs-core": "^2.19.16",
    "@bentley/frontend-application-insights-client": "^2.19.16",
    "@bentley/frontend-authorization-client": "^2.19.16",
    "@bentley/geometry-core": "^2.19.16",
    "@bentley/icons-generic": "^1.0.34",
    "@bentley/imodelhub-client": "^2.19.16",
    "@bentley/imodeljs-common": "^2.19.16",
    "@bentley/imodeljs-frontend": "^2.19.16",
    "@bentley/imodeljs-i18n": "^2.19.16",
    "@bentley/imodeljs-markup": "^2.19.16",
    "@bentley/imodeljs-quantity": "^2.19.16",
    "@bentley/itwin-client": "^2.19.16",
    "@bentley/orbitgt-core": "^2.19.16",
    "@bentley/presentation-common": "^2.19.16",
    "@bentley/presentation-frontend": "^2.19.16",
    "@bentley/product-settings-client": "^2.19.16",
    "@bentley/rbac-client": "^2.19.16",
    "@bentley/react-scripts": "^4.0.3",
    "@bentley/telemetry-client": "^2.19.16",
    "@bentley/ui-abstract": "^2.19.16",
    "@bentley/ui-components": "^2.19.16",
    "@bentley/ui-core": "^2.19.16",
    "@bentley/ui-framework": "^2.19.16",
    "@bentley/ui-ninezone": "^2.19.16",
    "@bentley/webgl-compatibility": "^2.19.16",
    "@itwin/web-viewer-react": "^1.3.0",
    "@types/history": "^4.7.9",
    "@types/node": "^14.17.26",
    "@types/react": "^16.14.17",
    "@types/react-dom": "^16.9.14",
    "history": "^4.10.1",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "typescript": "^4.3.5"
  },
  "scripts": {
    "start": "react-scripts --max_old_space_size=4096 start",
    "build": "react-scripts --max_old_space_size=4096 build",
    "test": "react-scripts --max_old_space_size=4096 test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": [
    "last 4 chrome version",
    "last 4 firefox version",
    "last 4 safari version",
    "last 4 ios version",
    "last 4 ChromeAndroid version",
    "last 4 edge version",
    "not dead",
    "not <0.2%"
  ]
}

**Reply by aruniverse** [reply]
That's because you didn't change the dependency on @bentley/react-scripts

```
"@bentley/react-scripts": "^4.0.3",
```

Change that to ^4.0.5 and do another  npm install.

**Comment by ksurell** [comment]
Sorry if I'm missing something but still same error.
This is what npm install says:

PS C:\EXTERN\prj\MSDev\iTwin\boo-1-483> npm install
npm WARN @babel/plugin-bugfix-v8-spread-parameters-in-optional-chaining@7.15.4 requires a peer of @babel/core@^7.13.0 but none is installed. You must install peer dependencies yourself.
npm WARN @bentley/webpack-tools-core@2.19.16 requires a peer of webpack@4.42.0 but none is installed. You must install peer dependencies yourself.
npm WARN @itwin/viewer-react@1.8.0 requires a peer of @bentley/imodeljs-backend@^2.14.0 but none is installed. You must install peer dependencies yourself.
npm WARN @pmmmwh/react-refresh-webpack-plugin@0.5.1 requires a peer of react-refresh@^0.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-i18next@11.3.3 requires a peer of i18next@>= 19.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.3.2 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.13 (node_modules\webpack-dev-server\node_modules\fsevents):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.13 (node_modules\watchpack-chokidar2\node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

**Reply by aruniverse** [reply]
These look like warnings. So should be ok. When you run `npm run start` what happens?

**Reply by ksurell** [reply]
I get the same exact error. 

My workflow is this:

npx create-react-app your-app-name --template @itwin/web-viewer --scripts-version @bentley/react-scripts
Change to "@bentley/react-scripts": "^4.0.5"
npm install
mspm run start

Failed to compile
multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js ./node_modules/react-dev-utils/webpackHotDevClient.js ./src/index.tsx
Module not found: Cannot find module 'C:\Users\krist\your-app-name\node_modules\react-refresh\index.js'. Please verify that the package.json has a valid "main" entry

**Reply by aruniverse** [reply]
Could you check your node_modules for that dependency on react-refresh and verify what version got pulled in? Should be 0.10.x

**Reply by ksurell** [reply]
I hope I understand you right.
This is the info in the file **package.json** in the folder **.\node_modules\react-refresh.**

{
  "_from": "react-refresh@^0.8.3",
  "_id": "react-refresh@0.8.3",
  "_inBundle": false,
  "_integrity": "sha512-X8jZHc7nCMjaCqoU+V2I0cOhNW+QMBwSUkeXnTi8IPe6zaRWfn60ZzvFDZqWPfmSJfjub7dDW1SP0jaHWLu/hg==",
  "_location": "/react-refresh",
  "_phantomChildren": {},
  "_requested": {
    "type": "range",
    "registry": true,
    "raw": "react-refresh@^0.8.3",
    "name": "react-refresh",
    "escapedName": "react-refresh",
    "rawSpec": "^0.8.3",
    "saveSpec": null,
    "fetchSpec": "^0.8.3"
  },
  "_requiredBy": [
    "/@bentley/react-scripts"
  ],
  "_resolved": "https://registry.npmjs.org/react-refresh/-/react-refresh-0.8.3.tgz",
  "_shasum": "721d4657672d400c5e3c75d063c4a85fb2d5d68f",
  "_spec": "react-refresh@^0.8.3",
  "_where": "C:\\EXTERN\\prj\\msdev\\iTwin\\boo-1-483\\node_modules\\@bentley\\react-scripts",
  "bugs": {
    "url": "https://github.com/facebook/react/issues"
  },
  "bundleDependencies": false,
  "deprecated": false,
  "description": "React is a JavaScript library for building user interfaces.",
  "engines": {
    "node": ">=0.10.0"
  },
  "files": [
    "LICENSE",
    "README.md",
    "babel.js",
    "runtime.js",
    "build-info.json",
    "cjs/",
    "umd/"
  ],
  "homepage": "https://reactjs.org/",
  "keywords": [
    "react"
  ],
  "license": "MIT",
  "main": "index.js",
  "name": "react-refresh",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/facebook/react.git",
    "directory": "packages/react"
  },
  "version": "0.8.3"
}

**Reply by aruniverse** [reply]
Ok something is wrong here because even though you've set @bentley/react-scripts to 4.0.5 its pulling in react-refresh@0.8.3 even though in @bentley/react-scripts@4.0.5 we tell it to pull react-refresh@^0.10.0, [see](https://github.com/imodeljs/create-react-app/blob/imodeljs/packages/react-scripts/package.json#L80).

Try deleting your node_modules and doing a fresh install; that should hopefully work.

**Reply by ksurell** [reply]
That did it!
Thanks for your patience.

/Krister
